crates/ika-core/src/dwallet_mpc/mpc_session.rs

```diff
@@ -1,1340 +1,294 @@
-                error!("failed to submit an MPC message to consensus: {:?}", err);
-            }
-        });
-        Ok(())
-    }
+        let mpc_round_messages_map = consensus_round_messages_map
+            .entry(message.round_number)
+            .or_default();
 
-    /// Report that the session failed because the threshold was not reached.
-    /// This is submitted to the consensus,
-    /// in order to make sure that all the Validators agree that this session needs more messages.
-    fn report_threshold_not_reached(&self, tokio_runtime_handle: &Handle) -> DwalletMPCResult<()> {
-        let report = ThresholdNotReachedReport {
-            session_identifier: self.session_identifier,
-            attempt: self.attempts_count,
-        };
-        let report_tx = self.new_dwallet_report_threshold_not_reached(report)?;
-        let epoch_store = self.epoch_store()?.clone();
-        let consensus_adapter = self.consensus_adapter.clone();
-        tokio_runtime_handle.spawn(async move {
-            if let Err(err) = consensus_adapter
-                .submit_to_consensus(&[report_tx], &epoch_store)
-                .await
-            {
-                error!(
-                    ?err,
-                    "failed to submit `threshold not reached` report to consensus"
-                );
-            }
-        });
-        Ok(())
+        if !mpc_round_messages_map.contains_key(&sender_party_id) {
+            mpc_round_messages_map.insert(sender_party_id, message.message);
+        }
     }
 
-    fn advance_specific_party(
+    /// This function iterates over the messages from different parties sent for different MPC rounds, ordered by the consensus round they were received.
+    ///
+    /// It builds a list of messages in order to advance the current round `current_mpc_round`,
+    /// using all the messages from the first consensus round to the first that satisfies the following conditions:
+    /// - a quorum of messages from the previous round `current_mpc_round - 1` must exist (except for the first round, which is always ready to advance requiring no messages as input.)
+    /// - a minimum number of consensus rounds that was required to delay the execution (in order to allow more messages to come in before advancing)
+    ///   has passed since the first consensus round where we got a quorum for this round.
+    /// - this quorum must be "fresh", in the sense we never tried to advance with it before.
+    ///   There is only one case in which we attempt to advance the same round twice: when we get a threshold not reached error.
+    ///   Therefore, if such an error occurred for a consensus round, we don't stop the search,
+    ///   and wait for at least one new message to come in a subsequent consensus round before returning the messages to advance with.
+    ///
+    /// Duplicate messages are ignored - the first message a party has sent for an MPC round is always used.
+    pub(crate) fn build_messages_to_advance(
```

Comment:
I feel like this function should not be from MPCSession but to receive the info you need from MPCSession and build it

Reply:
I also thought of that